lib/Sema/TypeCheckAttr.cpp

```diff
@@ -4581,6 +4583,39 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
   (void)attr->getParameterIndices();
 }
 
+/// Checks if original candidate and registered derivative match in terms
+/// of static declaration. If the original candidate is a constructor or is
+/// defined as static method then the registered derivative is expected to be static.
+/// Otherwise the registered derivative should be an instance method.
+/// \returns true if mismatch is found, false otherwise.
+static bool checkStaticDeclMismatch(AbstractFunctionDecl *originalCandidate,
+                                    AbstractFunctionDecl *registered) {
+  return (isa<ConstructorDecl>(originalCandidate) ||
+          originalCandidate->isStatic()) != registered->isStatic();
+}
+
+/// Produces diagnostics for mismatch in static/instance method declaration
+/// between original candidate and registered derivative.
+static void diagnoseStaticDeclMismatch(AbstractFunctionDecl *originalCandidate,
```

Comment:
I started defining these helper functions as you suggested as lambdas within `typeCheckDerivativeAttr` but the diagnostics apply as well for the `@transpose` attribute. Type-checking for `@transpose` attribute happens in `visitTransposeAttr` hence the definitions here so that we can reuse them in both places. I could define twice the a lambda that performs the checks as it is not much code. wdyt?

Reply:
I’m planning on revamping `@transpose` typing rules so the diagnostics may not apply to both anymore. So duplicating the core form now sounds good.

Reply:
I have pushed a new commit that duplicates the check and diagnostics for both attribute type check routines. As well, I have split the diagnostic messages for both attributes given that in one case one refers to the **derivative** function and to the **transpose** function in the other, plus it would be easier to remove them for the '@transpose' attribute if they need to go as a result of your refactoring.